grpc_unary_attestation/src/server.rs

```diff
@@ -18,20 +18,92 @@
 //! protocol.
 
 use crate::proto::{unary_session_server::UnarySession, UnaryRequest, UnaryResponse};
+use lru::LruCache;
+use oak_remote_attestation::handshaker::{AttestationBehavior, Encryptor, ServerHandshaker};
 use oak_utils::LogError;
+use std::sync::{Arc, Mutex};
 use tonic;
 
-/// gRPC Attestation Service implementation.
-pub struct AttestationServer<F, L: LogError> {
+type SessionId = u64;
+enum SessionState {
+    // Boxed due to large size difference, ref: https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
+    HandshakeInProgress(Box<ServerHandshaker>),
+    EncryptedMessageExchange(Encryptor),
+}
+
+/// Maintains remote attestation state for a number of sessions
+struct SessionsTracker {
     /// PEM encoded X.509 certificate that signs TEE firmware key.
-    _tee_certificate: Vec<u8>,
-    /// Processes data from client requests and creates responses.
-    _request_handler: F,
+    tee_certificate: Vec<u8>,
     /// Configuration information to provide to the client for the attestation step.
-    _additional_info: Vec<u8>,
+    additional_info: Vec<u8>,
+    known_sessions: LruCache<SessionId, SessionState>,
+}
+
+impl SessionsTracker {
+    pub fn create(tee_certificate: Vec<u8>, additional_info: Vec<u8>) -> Self {
+        let known_sessions = LruCache::new(10000);
+        Self {
+            tee_certificate,
+            additional_info,
+            known_sessions,
+        }
+    }
+
+    /// Consumes remote attestation state of an existing session. Creates
+    /// intial state if the session is not known.
+    ///
+    /// Note that getting the remote attestation state of a session always
+    /// implicitly removes it from the set of tracked sessions. After
+    /// using the state to process a request with this state it must explicitly
+    /// be put back into the SessionsTracker. This an intentional choice meant
+    /// to ensure that faulty state that leads to errors when processing
+    /// a request is not persistent.
+    pub fn pop_session_state(
+        &mut self,
+        session_id: SessionId,
+    ) -> anyhow::Result<SessionState, String> {
+        return match self.known_sessions.pop(&session_id) {
+            None => match AttestationBehavior::create_self_attestation(&self.tee_certificate) {
+                Ok(behavior) => Ok(SessionState::HandshakeInProgress(Box::new(
+                    ServerHandshaker::new(behavior, self.additional_info.clone()),
+                ))),
+                Err(error) => Err(format!(
```

Comment:
Consider https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.context (if it works here), and / or below.